execution_engine/src/core/engine_state/mod.rs

```diff
@@ -130,7 +130,24 @@ where
         ee_config: &ExecConfig,
     ) -> Result<GenesisResult, Error> {
         // Preliminaries
-        let initial_root_hash = self.state.empty_root();
+        let initial_root_hash = if let Some(state_root_hash) = ee_config.state_root_hash() {
+            let missing_descendants = self
+                .state
+                .missing_trie_keys(correlation_id, state_root_hash)
+                .map_err(|err| Error::Exec(err.into()))?;
+            if !missing_descendants.is_empty() {
+                return Err(Error::Genesis(GenesisError::MissingStateKeys {
+                    state_root_hash,
+                    missing_descendants,
+                }));
```

Comment:
@henrytill I paired with @fizyk20 and we are using this as an integrity check.

As @fizyk20 noted in our discussion, we still need to enforce the balances of the validators, but I believe that will be in a follow up PR.

Reply:
I think I'll have to do the stuff with balances in this PR, too - that is quite an important part of the functionality. I'm just in the process of figuring out how to do it correctly ;)